File: Data Structures & Algorithms/meeting-schedule/submission-6.java
/**
 * Definition of Interval:
 * public class Interval {
 *     public int start, end;
 *     public Interval(int start, int end) {
 *         this.start = start;
 *         this.end = end;
 *     }
 * }
 */

class Solution {
    public boolean canAttendMeetings(List<Interval> intervals) {
        // sort by start time
        // check if any intervals overlap
        Collections.sort(intervals, (a, b) -> a.start - b.start);
        // intervals.sort(Comparator.comparing(Interval::start));

        for (int i = 0; i < intervals.size() - 1; i++) {
            if (intervals.get(i + 1).start < intervals.get(i).end) {
                return false;
            }
        }

        return true;
    }
}

File: Data Structures & Algorithms/meeting-schedule-ii/submission-6.py
"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        if not intervals:
            return 0

        # sort by interval start
        intervals.sort(key=lambda interval: (interval.start, interval.end))

        res = 0
        prevEnd = -1
        visited = set()
        while len(visited) != len(intervals):
            for i in range(len(intervals)):
                if i not in visited:
                    if prevEnd == -1:
                        res += 1
                        prevEnd = intervals[i].end
                        visited.add(i)
                    else:
                        if intervals[i].start >= prevEnd:
                            prevEnd = intervals[i].end
                            visited.add(i)
            prevEnd = -1
        
        return res
